pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go

```diff
@@ -72,7 +83,67 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
 	return nil
 }
 
-// WithClient adds to the context a fake client with the desired behaviour
-func WithClient(ctx context.Context, behaviour *FakeClientBehaviour) context.Context {
-	return context.WithValue(ctx, ceKey{}, newFakeClient(behaviour))
+// addCount can be used to add the count when each event is going to be sent
+func (c FakeClient) addCount() {
+	c.waitGroup.Add(1)
+}
+
+// decreaseCount can be used to the decrease the count when each event is sent
+func (c FakeClient) decreaseCount() {
+	c.waitGroup.Done()
+}
+
+// WithClient adds to the context a fake client with the desired behaviour and expectedEventCount
+func WithClient(ctx context.Context, behaviour *FakeClientBehaviour, expectedEventCount int) context.Context {
+	return context.WithValue(ctx, ceKey{}, newFakeClient(behaviour, expectedEventCount))
+}
+
+// CheckCloudEventsUnordered checks that all events in wantEvents, and no others,
+// were received via the given chan, in any order.
+// Block until all events have been sent.
+// If more events were sent than expected, this will block indefinitely because the buffered channel
+// is initialized to the expected number of events and will not be able to send any more.
+func (c *FakeClient) CheckCloudEventsUnordered(t *testing.T, testName string, wantEvents []string) {
+	t.Helper()
+	c.waitGroup.Wait()
+	err := eventsFromChannelUnordered(c.Events, wantEvents)
+	if err != nil {
+		t.Fatalf("error getting events from fake client: %v", err)
+	}
+}
+
+// eventsFromChannelUnordered takes a chan of string and a list of events that a test
+// expects to receive. The events can be received in any order. Any extra or too few
+// events are both considered errors.
+func eventsFromChannelUnordered(c chan string, wantEvents []string) error {
+	expected := append([]string{}, wantEvents...)
+	channelEvents := len(c)
+	// fakeclient's channel buffersize equals to the size of wantEvents, so no extra events will be sent
```

Comment:
This comment is slightly misleading. Buffered channels can send once events have been pulled off of them. The reason no extra events will be sent is because the fake client returns an error in this case, but that logic lives far away from here. 

I'd suggest adding commentary to the fake client where you return an error (to explain why this prevents extra events from being sent) and some commentary here about why this block detects the case where too few events are sent.

Reply:
👍